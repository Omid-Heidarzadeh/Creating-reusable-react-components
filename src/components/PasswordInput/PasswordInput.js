import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import TextInput from 'ps-react/TextInput';
import EyeIcon from 'ps-react/EyeIcon';
import ProgressBar from 'ps-react/ProgressBar';

function PasswordInput({
  label = 'Password',
  htmlFor = 'password',
  value,
  minLength = 8,
  maxLength = 20,
  error = '',
  onChange,
  quality = 0,
  placeholder,
  showVisibilityToggle = true,
  style,
  ...props
}) {
  const [showPassword, setShowPassword] = useState(false);
  const eyeContainerRef = useRef(null);
  const inputRef = useRef(null);
  const type = showPassword ? 'text' : 'password';

  useEffect(() => {
    if (!showVisibilityToggle) return;

    const eyeIcon = eyeContainerRef.current;
    const containerRect = eyeContainerRef.current.getBoundingClientRect();
    const parentRect =
      eyeContainerRef.current.parentElement.getBoundingClientRect();
    const inputRect = inputRef.current.getBoundingClientRect();

    eyeIcon.style.left =
      inputRect.left - parentRect.left + inputRect.width - 20 + 'px';
    eyeIcon.style.top =
      inputRect.top -
      parentRect.top +
      (inputRect.height - containerRect.height) / 2 +
      'px';
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div
      style={{
        position: 'relative',
        marginBottom: '1em',
        ...style?.container,
      }}
    >
      <TextInput
        label={label}
        htmlFor={htmlFor}
        type={type}
        value={value}
        minLength={minLength}
        maxLength={maxLength}
        placeholder={placeholder || 'Enter password'}
        onChange={onChange}
        error={error}
        inputRef={inputRef}
        required
        style={{
          label: { ...style?.label },
          input: {
            ...style?.input,
            paddingRight: '1.5em',
            paddingLeft: '.5em',
          },
        }}
        {...props}
      >
        <ProgressBar style={{ marginTop: '.5em' }} percent={quality} />
      </TextInput>
      {showVisibilityToggle && (
        <span
          style={{
            position: 'absolute',
            display: 'inline-flex',
          }}
          ref={eyeContainerRef}
          onClick={() => setShowPassword((showPass) => !showPass)}
        >
          <EyeIcon />
        </span>
      )}
    </div>
  );
}

PasswordInput.propTypes = {
  /** Label text */
  label: PropTypes.string,

  /** Input field ID */
  htmlFor: PropTypes.string,

  /** Input field value */
  value: PropTypes.string.isRequired,

  /** Input field value min-length */
  minLength: PropTypes.string,

  /** Input field value max-length */
  maxLength: PropTypes.string,

  /** Input field placeholder */
  placeholder: PropTypes.string,

  /** Error message */
  error: PropTypes.string,

  /** Input field change event handler */
  onChange: PropTypes.func.isRequired,

  /** Computed password quality in percent */
  quality: PropTypes.number,

  /** Show visibility toggle icon */
  showVisibilityToggle: PropTypes.bool,

  /** Customized styles */
  style: PropTypes.shape({
    label: PropTypes.object,
    input: PropTypes.object,
  }),
};

export default PasswordInput;
