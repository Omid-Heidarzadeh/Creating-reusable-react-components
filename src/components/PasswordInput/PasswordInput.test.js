import React from 'react';
import Renderer from 'react-test-renderer';
import PasswordInput from './PasswordInput';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

configure({ adapter: new Adapter() });

describe('password input', () => {
  it('should render a password input without quality bar (using Enzyme)', () => {
    const result = shallow(
      <PasswordInput
        label="Password"
        htmlFor="password"
        value="gj32o5hasd"
        onChange={jest.fn()}
        quality={50}
        showQuality={false}
      />
    );
    expect(result.html()).toMatchSnapshot();
  });

  it('should render a password input without quality bar (using test-renderer)', () => {
    const result = Renderer.create(
      <PasswordInput
        label="Password"
        htmlFor="password"
        value="gj32o5hasd"
        onChange={jest.fn()}
        quality={50}
        showQuality={false}
      />
    ).toJSON();
    console.log(result);
    expect(result).toMatchSnapshot();
  });

  it('should toggle input type when eyeicon is clicked', () => {
    const component = mount(
      <PasswordInput
        label="Password"
        htmlFor="password"
        value="gj32o5hasd"
        onChange={jest.fn()}
        quality={50}
        showQuality={false}
      />
    );
    /** Default input type should be password */
    expect(component.find('input').getElement().props.type).toEqual('password');

    /** Input type should toggle after click on toggle icon */
    component.find('.password-visibility-toggle').simulate('click');
    expect(component.find('input').getElement().props.type).toEqual('text');

    /** Input type should toggle after click on toggle icon */
    component.find('.password-visibility-toggle').simulate('click');
    expect(component.find('input').getElement().props.type).toEqual('password');
  });
});
