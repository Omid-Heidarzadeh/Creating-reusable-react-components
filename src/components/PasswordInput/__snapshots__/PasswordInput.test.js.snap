// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`password input should render a password input without quality bar (using Enzyme) 1`] = `"<div style=\\"position:relative;margin-bottom:1em\\"><div style=\\"margin-bottom:1em\\"><label for=\\"password\\" style=\\"display:block;margin-bottom:.5em\\">Password<span style=\\"color:red\\">*</span></label><span style=\\"position:relative;display:inline-block;vertical-align:middle\\"><input id=\\"password\\" type=\\"password\\" placeholder=\\"Enter password\\" minLength=\\"8\\" maxLength=\\"20\\" required=\\"\\" style=\\"line-height:1.5;padding-right:1.5em;padding-left:.5em\\" autoComplete=\\"current-password\\" value=\\"gj32o5hasd\\"/></span></div><span class=\\"password-visibility-toggle\\" style=\\"position:absolute;display:inline-flex\\"><svg width=\\"16\\" height=\\"16\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 22 22\\"><g transform=\\"matrix(.02146 0 0 .02146 1 1)\\" fill=\\"#4d4d4d\\"><path d=\\"m466.07 161.53c-205.6 0-382.8 121.2-464.2 296.1-2.5 5.3-2.5 11.5 0 16.9 81.4 174.9 258.6 296.1 464.2 296.1 205.6 0 382.8-121.2 464.2-296.1 2.5-5.3 2.5-11.5 0-16.9-81.4-174.9-258.6-296.1-464.2-296.1m0 514.7c-116.1 0-210.1-94.1-210.1-210.1 0-116.1 94.1-210.1 210.1-210.1 116.1 0 210.1 94.1 210.1 210.1 0 116-94.1 210.1-210.1 210.1\\"></path><circle cx=\\"466.08\\" cy=\\"466.02\\" r=\\"134.5\\"></circle></g></svg></span></div>"`;

exports[`password input should render a password input without quality bar (using test-renderer) 1`] = `
<div
  style={
    Object {
      "marginBottom": "1em",
      "position": "relative",
    }
  }
>
  <div
    style={
      Object {
        "marginBottom": "1em",
      }
    }
  >
    <label
      htmlFor="password"
      style={
        Object {
          "color": "",
          "display": "block",
          "marginBottom": ".5em",
        }
      }
    >
      Password
      <span
        style={
          Object {
            "color": "red",
          }
        }
      >
        *
      </span>
    </label>
    <span
      style={
        Object {
          "display": "inline-block",
          "position": "relative",
          "verticalAlign": "middle",
        }
      }
    >
      <input
        autoComplete="current-password"
        id="password"
        maxLength="20"
        minLength="8"
        onBlur={[Function]}
        onChange={[MockFunction]}
        placeholder="Enter password"
        required={true}
        style={
          Object {
            "borderColor": "",
            "lineHeight": 1.5,
            "paddingLeft": ".5em",
            "paddingRight": "1.5em",
          }
        }
        type="password"
        value="gj32o5hasd"
      />
    </span>
  </div>
  <span
    className="password-visibility-toggle"
    onClick={[Function]}
    style={
      Object {
        "display": "inline-flex",
        "position": "absolute",
      }
    }
  >
    <svg
      height="16"
      viewBox="0 0 22 22"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="#4d4d4d"
        transform="matrix(.02146 0 0 .02146 1 1)"
      >
        <path
          d="m466.07 161.53c-205.6 0-382.8 121.2-464.2 296.1-2.5 5.3-2.5 11.5 0 16.9 81.4 174.9 258.6 296.1 464.2 296.1 205.6 0 382.8-121.2 464.2-296.1 2.5-5.3 2.5-11.5 0-16.9-81.4-174.9-258.6-296.1-464.2-296.1m0 514.7c-116.1 0-210.1-94.1-210.1-210.1 0-116.1 94.1-210.1 210.1-210.1 116.1 0 210.1 94.1 210.1 210.1 0 116-94.1 210.1-210.1 210.1"
        />
        <circle
          cx="466.08"
          cy="466.02"
          r="134.5"
        />
      </g>
    </svg>
  </span>
</div>
`;
